lib/aws.go

```diff
@@ -0,0 +1,345 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package lib
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	v4_creds "github.com/aws/aws-sdk-go-v2/aws/signer/v4" // ¯\_(ツ)_/¯
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+)
+
+// AWS returns a cel.EnvOption to configure extended functions for
+// AWS request signing. In all cases the functions are member overloads
+// on a <map<string,dyn>> which is expected to correspond to an HTTP
+// request, the returned value is the request with the necessary AWS v4
+// signing details added.
+//
+// # Sign AWS from env
+//
+// Returns a signed request based on credential details in the relevant
+// AWS environment variables; AWS_ACCESS_KEY, AWS_ACCESS_KEY_ID,
+// AWS_SECRET_ACCESS_KEY and AWS_SECRET_KEY.
```

Comment:
Should we expand this list to cover more of what the SDK supports? Or just state that it supports the AWS SDK env vars without calling them out specifically (aws publishes a list at https://docs.aws.amazon.com/sdkref/latest/guide/settings-reference.html#EVarSettings).

```suggestion
// Returns a signed request based on credential details in the relevant
// AWS environment variables. Supports AWS SDK v2 environment variables including:
//
// Credentials:
//   - AWS_ACCESS_KEY_ID (primary) or AWS_ACCESS_KEY (fallback)
//   - AWS_SECRET_ACCESS_KEY (primary) or AWS_SECRET_KEY (fallback)
//   - AWS_SESSION_TOKEN (optional)
//
// Region:
//   - AWS_REGION or AWS_DEFAULT_REGION
//
// Profile and configuration files:
//   - AWS_PROFILE or AWS_DEFAULT_PROFILE
//   - AWS_SHARED_CREDENTIALS_FILE
//   - AWS_CONFIG_FILE
//
// Other supported variables:
//   - AWS_WEB_IDENTITY_TOKEN_FILE
//   - AWS_ROLE_ARN
//   - AWS_ROLE_SESSION_NAME
//   - AWS_CA_BUNDLE
```

Reply:
I will link to the docs and remove the explicit list.